### PR TITLE
[VGNWB-209] added functionality to delete relation templates

### PR DESCRIPTION
**Pull request by prnvshrn**

### CODE REVIEW

File: annotations/templates/annotations/relationtemplate_list.html [hunk 1]
```diff
@@ -26,16 +26,61 @@
...
+                          </div>
+                          <div class="modal-footer">
+                            <input class="btn btn-danger" type="submit" value="Delete">
+                            <button type="button" class="btn btn-default" data-dismiss="modal">Close</button>
+                          </div>
+                          </form>
+                        </div>
+                      </div>
+                    </div>
+                    <!-- Delete Template modal section ends here -->
```
Comment (jdamerow) [inline annotations/templates/annotations/relationtemplate_list.html]: I wouldn't create a modal for every single template. That just becomes unmanageable when there are many templates. Rather have one modal for which you set the selected template id when the delete button is clicked.

File: annotations/templates/annotations/relationtemplate_list.html [hunk 2]
```diff
@@ -46,9 +91,21 @@
...
 <script>
 jQuery(document).ready(function($) {
+
     $(".clickable-row").click(function() {
         window.document.location = $(this).data("href");
     });
+
+    $('[data-delete-template]').tooltip();
+
+    $("a[id^='openTemplateDeleteModal']").click(function(e) {
```
Comment (jdamerow) [inline annotations/templates/annotations/relationtemplate_list.html]: just give all buttons the same class and then use the class to identify buttons here.

File: annotations/views/relationtemplate_views.py [hunk 3]
```diff
@@ -208,5 +206,12 @@ def create_from_relationtemplate(request, template_id):
     else:   # Not sure if we want to do anything for GET requests at this point.
         response_data = {}
 
-
     return JsonResponse(response_data)
+
+
+def delete_relationtemplate(request, template_id):
```
Comment (jdamerow) [inline annotations/views/relationtemplate_views.py]: this needs to be access restricted.

**Review by jdamerow — CHANGES_REQUESTED** [review]

File: annotations/templates/annotations/relationtemplate_list.html (hunk 1 above)
Comment (prnvshrn) [reply annotations/templates/annotations/relationtemplate_list.html]: Changed the implementation of creating multiple modals to a single modal.

File: annotations/templates/annotations/relationtemplate_list.html (hunk 2 above)
Comment (prnvshrn) [reply annotations/templates/annotations/relationtemplate_list.html]: The buttons now are identified by class.

File: annotations/views/relationtemplate_views.py (hunk 3 above)
Comment (prnvshrn) [reply annotations/views/relationtemplate_views.py]: Added "staff_member_required" before <b>delete_relationtemplate</b> method to restrict access.

**Comment by prnvshrn** [comment]
I have made the following changes:-
1. Changed multiple modals to a single modal
2. Using class to identify buttons for opening the modal
3. Added 'staff_member_required' to restrict access

File: annotations/views/relationtemplate_views.py [hunk 4]
```diff
@@ -208,5 +206,13 @@ def create_from_relationtemplate(request, template_id):
...
 
-
     return JsonResponse(response_data)
+
+
+@staff_member_required
+def delete_relationtemplate(request, template_id):
+    if request.method == 'POST':
+        RelationTemplate.objects.filter(id=template_id).delete()
+        RelationTemplatePart.objects.filter(id=template_id).delete()
```
Comment (jdamerow) [inline annotations/views/relationtemplate_views.py]: this should be done in a transaction:
https://docs.djangoproject.com/en/1.11/topics/db/transactions/#controlling-transactions-explicitly

**Review by jdamerow — CHANGES_REQUESTED** [review]

File: annotations/views/relationtemplate_views.py (hunk 4 above)
Comment (prnvshrn) [reply annotations/views/relationtemplate_views.py]: Changed implementation of deletion logic to execute in an atomic transaction

**Comment by prnvshrn** [comment]
Changed implementation of deletion logic to execute in an atomic transaction

**Comment by jdamerow** [comment]
something is going wrong here. try this: 
- create 3 templates (two simple subject-predicate-object ones, and the last one nested, not sure if this is important, but that's what I did). 
- then delete the last one (the nested one).
- no look at the first one: -> relation parts are deleted

**Comment by jdamerow** [comment]
also, it looks like relationsets have references to relation templates, which means relation templates are deleted that are referenced by relationsets, there will most likely be unintended side effects. A relation template should probably not be allowed to be deleted if it is used in a relation set.

**Comment by prnvshrn** [comment]
Added the following changes -
1. Wrong deletion of relation template part was occurring because in the views I was passing <b>id</b> instead of <b>part_of</b> for deletion. So I changed the method to use <b>part_of</b> instead.
2. Added check to see if data is present in relationset for a template before deletion. If data is present, an error is shown and template is not deleted.

**Review by jdamerow — APPROVED** [review]

**Pull request merged**